File: excelium-common/src/main/java/excelium/common/PlatformDetector.java
/*
 * MIT License
 *
 * Copyright (c) 2018 Excelium
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package excelium.common;

import excelium.model.enums.Platform;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.Locale;

/**
 * Platform detector
 *
 * @author PhungDucKien
 * @since 2018.04.17
 */
public class PlatformDetector {

    /**
     * Logger
     */
    private static final Logger LOG = LogManager.getLogger();

    /**
     * Detected platform. For later uses
     */
    private static Platform detectedPlatform;

    /**
     * Gets system platform
     *
     * @return system platform
     */
    public static Platform getPlatform() {
        if (detectedPlatform == null) {
            String OS = getOsName().toLowerCase(Locale.ENGLISH);
            if (OS.contains("mac") || OS.contains("darwin")) {
                detectedPlatform = Platform.MAC_OS;
            } else if (OS.contains("win")) {
                String arch = getOsArch().toLowerCase(Locale.ENGLISH);
                if (arch.contains("64")) {
                    detectedPlatform = Platform.WINDOWS_64;
                } else {
                    detectedPlatform = Platform.WINDOWS_32;
                }
            } else if (OS.contains("nux")) {
                detectedPlatform = Platform.LINUX;
            } else {
                detectedPlatform = Platform.OTHER;
            }
            LOG.info("Detected Platform: {}", detectedPlatform.getText());
        }
        return detectedPlatform;
    }

    /**
     * Gets OS name
     *
     * @return OS name
     */
    private static String getOsName() {
        return System.getProperty("os.name", "generic");
    }

    /**
     * Gets OS arch
     *
     * @return OS arch
     */
    private static String getOsArch() {
        return System.getProperty("os.arch");
    }
}

File: packages/common/src/PlatformDetector.ts
/*
 * Licensed to the Software Freedom Conservancy (SFC) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The SFC licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Platform } from '@excelium/model';

/**
 * Platform detector
 *
 * @author PhungDucKien
 * @since 2018.04.17
 */
export default class PlatformDetector {
  /**
   * Gets system platform
   *
   * @return system platform
   */
  public static getPlatform(): Platform {
    if (PlatformDetector.detectedPlatform == null) {
      const OS = PlatformDetector.getOsName().toLowerCase();
      if (OS.indexOf('mac') > -1 || OS.indexOf('darwin') > -1) {
        PlatformDetector.detectedPlatform = Platform.MAC_OS;
      } else if (OS.indexOf('win') > -1) {
        const arch = PlatformDetector.getOsArch().toLowerCase();
        if (arch.indexOf('64') > -1) {
          PlatformDetector.detectedPlatform = Platform.WINDOWS_64;
        } else {
          PlatformDetector.detectedPlatform = Platform.WINDOWS_32;
        }
      } else if (OS.indexOf('nux') > -1) {
        PlatformDetector.detectedPlatform = Platform.LINUX;
      } else {
        PlatformDetector.detectedPlatform = Platform.OTHER;
      }
    }
    return PlatformDetector.detectedPlatform;
  }

  /**
   * Gets OS name
   *
   * @return OS name
   */
  public static getOsName(): string {
    return process.platform;
  }

  /**
   * Gets OS arch
   *
   * @return OS arch
   */
  public static getOsArch(): string {
    return process.arch;
  }

  /**
   * Detected platform. For later uses
   */
  private static detectedPlatform: Platform;
}
